src/controller/python/chip/server/ServerInit.cpp

```diff
@@ -101,6 +101,13 @@ static bool EnsureWiFiIsStarted()
 }
 #endif
 
+void CleanShutdown()
+{
+    chip::DeviceLayer::PlatformMgr().LockChipStack();
```

Comment:
Why are we doing this?

Importantly, you should be shutting down the thread that's running the server by calling PlatformMgr().StopEventLoopTask() before doing any of the remaining shutdown bits.

Reply:
Without the call to `chip::DeviceLayer::PlatformMgr().LockChipStack();` the follow crash occurs:
```
[Thread 0x7fe67fc1c0 (LWP 50338) exited]

Thread 1 "python3" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x0000007ff7cfeaa0 in __GI_abort () at abort.c:79
#2  0x0000007ff5f60564 in chipAbort() () at ../../src/lib/support/CodeUtils.h:506
#3  0x0000007ff60a7044 in chipDie() () at ../../src/lib/support/CodeUtils.h:516
#4  0x0000007ff60a70b4 in chip::Platform::Internal::AssertChipStackLockedByCurrentThread(char const*, int) (file=0x7ff62a2560 "../../src/transport/Session.h", line=187)
    at ../../src/platform/LockTracker.cpp:36
#5  0x0000007ff60ba0c8 in chip::Transport::Session::RemoveHolder(chip::SessionHolder&) (this=0xb83c00, holder=...) at ../../src/transport/Session.h:187
#6  0x0000007ff60ba6a4 in chip::SessionHolder::Release() (this=0x7ff62fe968 <chip::Server::sServer+536>) at ../../src/transport/SessionHolder.cpp:113
#7  0x0000007ff625bdf4 in chip::PairingSession::Clear() (this=0x7ff62fe958 <chip::Server::sServer+520>) at ../../src/protocols/secure_channel/PairingSession.cpp:167
#8  0x0000007ff624848c in chip::CASESession::Clear() (this=0x7ff62fe938 <chip::Server::sServer+488>) at ../../src/protocols/secure_channel/CASESession.cpp:147
#9  0x0000007ff606e12c in chip::CASEServer::Shutdown() (this=0x7ff62fe8f8 <chip::Server::sServer+424>) at ../../src/protocols/secure_channel/CASEServer.h:50
#10 0x0000007ff6070bd8 in chip::Server::Shutdown() (this=Python Exception <class 'RecursionError'> maximum recursion depth exceeded while getting the str of an object: 
0x7ff62fe750 <chip::Server::sServer>) at ../../src/app/server/Server.cpp:420
#11 0x0000007ff5f65e00 in CleanShutdown() () at ../../src/controller/python/chip/server/ServerInit.cpp:108
#12 0x0000007ff7d1478c in __run_exit_handlers (status=0, listp=0x7ff7e4e6b8 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at exit.c:108
#13 0x0000007ff7d1491c in __GI_exit (status=<optimized out>) at exit.c:139
#14 0x0000007ff7cfee1c in __libc_start_main (main=
    0x587538 <_start+56>, argc=2, argv=0x7ffffff468, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=<optimized out>) at ../csu/libc-start.c:342
#15 0x0000000000587534 in _start ()

```

I'll add a call to `chip::Server::GetInstance().DispatchShutDownAndStopEventLoop();` to stop the event loop.

Reply:
That's because `DispatchShutDownAndStopEventLoop` just schedules the shutdown of the stack on the Matter thread, but it doesn't actually wait for it to execute.

So without grabbing the lock, it's attempting to shutdown the Server object from the wrong thread, because the Matter context is still running.

You need to make a couple of changes (checkout [this](https://github.com/project-chip/connectedhomeip/pull/20645/files#diff-6dae7c59ddec614a77fba8543d60970b6a9e37b72e9a87ed6cc0bc4c7cd437cbR124) PR for the basis for this):
1. Don't just pthread_create a new Server thread. There are methods to do this already, i.e `StartEventLoopTask`.
2. Consequently here, you should do the following then:
```
chip::DeviceLayer::PlatformMgr().StopEventLoopTask();
chip::Server::GetInstance().Shutdown();
chip::DeviceLayer::PlatformMgr().Shutdown();
```

That ensures a clean shutdown of the Matter thread, waits for it to complete, then shuts down the remaining resident objects.

Reply:
@mrjerryjohns thanks for the feedback, I've main the changes you suggested.